src/segmentation.cpp

```diff
@@ -141,36 +162,59 @@ bool GetSceneAboveSurface(pcl::PointCloud<pcl::PointXYZRGB>::Ptr cloud,
 }
 
 bool FindObjectsOnSurface(PointCloudC::Ptr cloud, pcl::PointIndicesPtr indices,
-                          const Surface& surface, double margin_above_surface,
-                          double cluster_distance, int min_cluster_size,
-                          int max_cluster_size,
-                          SurfaceObjects* surface_objects) {
-  pcl::PointIndices::Ptr above_surface_indices(new pcl::PointIndices());
-  bool success =
-      GetSceneAboveSurface(cloud, indices, *surface.coefficients,
-                           margin_above_surface, above_surface_indices);
-  if (!success) {
-    return false;
+                          const std::vector<Surface>& surface_vec,
+                          double margin_above_surface, double cluster_distance,
+                          int min_cluster_size, int max_cluster_size,
+                          std::vector<SurfaceObjects>* surfaces_objects_vec) {
+  // Copy the vector and sort by height
+  std::vector<Surface> surfaces = surface_vec;
+  std::sort(surfaces.begin(), surfaces.end(), SurfaceComparator);
+  std::vector<pcl::PointIndices::Ptr> above_surface_indices_vec;
+
+  // Obtain indices between each horizontal surfaces
+  for (size_t i = 0; i < surfaces.size(); i++) {
+    float height_limit = std::numeric_limits<float>::max();
+    pcl::PointIndices::Ptr above_surface_indices(new pcl::PointIndices);
+    if (i != (surfaces.size() - 1)) {
+      height_limit = -1.0 * surfaces[i + 1].coefficients->values[3] /
+                     surfaces[i + 1].coefficients->values[2];
+    }
+
+    bool success = GetSceneAboveSurface(
+        cloud, indices, *surfaces[i].coefficients, margin_above_surface,
+        height_limit, above_surface_indices);
+    if (!success) {
+      ROS_ERROR(
+          "Error: extraction of indices above a horizontal surface failed");
+      return false;
+    }
+
+    above_surface_indices_vec.push_back(above_surface_indices);
   }
 
-  std::vector<pcl::PointIndices> object_indices;
-  pcl::EuclideanClusterExtraction<PointC> euclid;
-  euclid.setInputCloud(cloud);
-  euclid.setIndices(above_surface_indices);
-  euclid.setClusterTolerance(cluster_distance);
-  euclid.setMinClusterSize(min_cluster_size);
-  euclid.setMaxClusterSize(max_cluster_size);
-  euclid.extract(object_indices);
-
-  surface_objects->surface = surface;
-  for (size_t i = 0; i < object_indices.size(); ++i) {
-    Object object;
-    object.cloud = cloud;
-    object.indices.reset(new pcl::PointIndices(object_indices[i]));
-    object.pose_stamped.header.frame_id = cloud->header.frame_id;
-    FitBox(cloud, object.indices, surface.coefficients,
-           &object.pose_stamped.pose, &object.dimensions);
-    surface_objects->objects.push_back(object);
+  for (size_t i = 0;
+       i < surfaces.size() && i < above_surface_indices_vec.size(); i++) {
```

Comment:
`surfaces` and `above_surface_indices_vec` are always going to be the same size, right? So you don't need to check the bounds on both?

Reply:
Completed